File: AoC/2023_problems/day15.py
test_data = "rn=1,cm-,qp=3,cm=2,qp-,pc=4,ot=9,ab=5,pc-,pc=6,ot=7"

with open("AoC/2023_problems/day15.txt") as f:
    data = f.read()

# Alg 0 -> + ASCII (char) -> *= 17 -> %= 256
def hash_alg(seq):
    hash = 0
    for s in seq:
        hash = ((hash + ord(s)) * 17) % 256
    return hash


def part1(data):
    instructions = data.split(",")
    hash = 0
    for ins in instructions:
        hash += hash_alg(ins)
    return hash


def part2(data):
    instructions = data.split(",")
    boxes = {k: dict() for k in range(256)}
    for ins in instructions:
        if "=" in ins:
            letters, focal_len = ins.split("=")
            box_num = hash_alg(letters)
            boxes[box_num][letters] = int(focal_len)
        else:
            letters = ins.split("-")[0]
            box_num = hash_alg(letters)
            if letters in boxes[box_num]:
                del boxes[box_num][letters]
    total_focus = 0
    for box in boxes:
        box_mul = box + 1
        for i, lens in enumerate(boxes[box].values()):
            total_focus += box_mul * (i+1) * lens
    return total_focus


print(part1(data))
print(part2(data))

File: AoC/2023_problems/day15.js
const fs = require("fs");

const data = fs.readFileSync("AoC/2023_problems/day15.txt", "utf8");

const hash = (text) => text.split("").reduce((acc, char) => ((acc + char.charCodeAt()) * 17) % 256, 0);

const part1 = (data) => data.split(",").reduce((acc, line) => (acc + hash(line)), 0);

function part2(data) {
    const lines = data.split(",");
    const boxes = Array.from({length: 256}, (x) => ({}));
    for (const line of lines) {
        if (line.includes("=")) {
            const [key, value] = line.split("=");
            boxes[hash(key)][key] = value;
        }
        else {
            const key = line.split("-")[0];
            delete boxes[hash(key)][key];
        }
    }
    const boxValue = (box, boxMulti) => Object.values(box).reduce((acc, val, index) => (acc + boxMulti * val * (index+1)), 0);
    return boxes.reduce((acc, box) => (acc + boxValue(box, boxes.indexOf(box)+1)), 0);
}

console.log(part1(data));
console.log(part2(data))
